vmm/src/device_manager.rs

```diff
@@ -980,20 +1049,38 @@ impl DeviceManager {
         pci: &mut PciBus,
         buses: &mut BusInfo,
         mem_slots: u32,
-    ) -> DeviceManagerResult<()> {
+    ) -> DeviceManagerResult<BTreeMap<u32, Arc<dyn ExternalDmaMapping>>> {
         let mut mem_slot = mem_slots;
+        let mut iommu_attached_list = BTreeMap::new();
         if let Some(device_list_cfg) = &vm_info.vm_cfg.devices {
             // Create the KVM VFIO device
             let device_fd = DeviceManager::create_kvm_device(vm_info.vm_fd)?;
             let device_fd = Arc::new(device_fd);
 
             for device_cfg in device_list_cfg.iter() {
+                // We need to shift the device id since the 3 first bits
+                // are dedicated to the PCI function, and we know we don't
+                // do multifunction. Also, because we only support one PCI
+                // bus, the bus 0, we don't need to add anything to the
+                // global device ID.
+                let device_id = pci.next_device_id() << 3;
+
                 let vfio_device =
                     VfioDevice::new(&device_cfg.path, device_fd.clone(), vm_info.memory.clone())
                         .map_err(DeviceManagerError::VfioCreate)?;
 
-                let mut vfio_pci_device = VfioPciDevice::new(vm_info.vm_fd, allocator, vfio_device)
-                    .map_err(DeviceManagerError::VfioPciCreate)?;
+                if device_cfg.iommu {
+                    let vfio_mapping: Arc<dyn ExternalDmaMapping> = Arc::new(VfioDmaMapping::new(
+                        vfio_device.get_container(),
+                        Arc::clone(vm_info.memory),
+                    ));
```

Comment:
If `VfioDevice` implements the `ExternalDmaMapping` trait, you won't need to build what's essentially a copy of your VFIO container and a reference to your guest memory.

Reply:
True, but as explained [here](https://github.com/intel/cloud-hypervisor/pull/325/files#r333794182), this would force us using Arc<Mutex<>> for the entire object, and this might have some performance impact.